File: backend/src/main/java/com/project/manager/services/AttachementService.java
package com.project.manager.services;

import com.project.manager.models.Attachement;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
@CrossOrigin("*")

public interface AttachementService {
        Attachement saveAttachement(MultipartFile file, Long taskId) throws Exception;
        Attachement getAttachement(Long fileId) throws Exception;
        void deleteAttachementByTaskId(Long taskId) throws Exception;
        Attachement updateAttachement(Long fileId, MultipartFile file, Long taskId) throws Exception;
        List<Attachement> getAllAttachements() throws Exception;
        Attachement getAttachementByTask(Long taskId) throws Exception;
}

File: frontend/src/app/services/attachement.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse, HttpEvent, HttpEventType } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { Attachement } from '../models/attachementModel';

@Injectable({
  providedIn: 'root'
})
export class AttachementService {
  private baseUrl = 'http://localhost:8080';

  constructor(private http: HttpClient) { }

  saveAttachement(file: File, taskId: number): Observable<any> {
    const formData: FormData = new FormData();
    formData.append('file', file);
    formData.append('taskId', taskId.toString());

    return this.http.post<any>(`${this.baseUrl}/uploadFile`, formData, {
      reportProgress: true,
      observe: 'events'
    })
  }

  getAllAttachements(): Observable<Attachement[]> {
    return this.http.get<Attachement[]>(`${this.baseUrl}/getAllAttachements`).pipe(
    );
  }

  getAttachementByTask(taskId: number): Observable<Attachement> {
    return this.http.get<Attachement>(`${this.baseUrl}/getAttachementByTask/${taskId}`).pipe(
    );
  }

  getAttachement(fileId: number): Observable<Blob> {
    return this.http.get(`${this.baseUrl}/downloadFile/${fileId}`, {
      responseType: 'blob',
    });
  }

  deleteAttachement(taskId: number): Observable<any> {
    return this.http.delete<any>(`${this.baseUrl}/deleteAttachement/${taskId}`).pipe(
    );
  }


  updateAttachement(fileId: number, file: File, taskId: number): Observable<Attachement> {
    const formData: FormData = new FormData();
    formData.append('file', file);
    formData.append('taskId', taskId.toString());

    return this.http.put<Attachement>(
      `${this.baseUrl}/updateFile/${fileId}`,
      formData
    )
  }
}

